Expand join pushdown documentation.

posulliv opened the pull request.

hashhar commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,65 @@ to perform the remaining query processing on a smaller amount of data.
+
+.. code-block:: text
+
+ Fragment 0 [SINGLE]
```
For the purposes of this doc maybe we can make use of `EXPLAIN (TYPE LOGICAL, FORMAT TEXT)` to get more condensed plans.

That way you can maybe also add a plan with pushdown disabled to contrast it against.

No strong opinions on this one, it's pretty good as is.

posulliv replied:
I prefer to go with `EXPLAIN (TYPE DISTRIBUTED)` simply because whenever I work with users, that is what they always use and are more familiar with since it's the default.

hashhar replied:
Agreed on the familiarity part. It'd be confusing for most users.

hashhar commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,65 @@ to perform the remaining query processing on a smaller amount of data.
+
+  Care should be taken when enabling join pushdown in a connector without statistics as
+  pushing down a join which can increase the row count compared to the size of
+  the input tables to the join may incur large costs both in terms of performance
```
nit: 
```suggestion
  the input to the join may incur large costs both in terms of performance
```

the reason for suggestion is that even if you have a 10GB table on both sides with 1000 columns but you only join 1 column across both of them with maybe 1MB of data then only that 1MB is the size to be concerned about. 

In other words it doesn't matter what the size of the table is - it matters what the size of the inputs to the join is. In other docs this is referred to as "left side" and "right side" of the join.

hashhar commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,65 @@ to perform the remaining query processing on a smaller amount of data.
+to be pushed down:
+
+* join pushdown must be enabled
+* all predicates that are part of the join must be possible to be pushed down
```
nit: although this point implicitly covers it this is such a common question that maybe it deserves a sub-point

```suggestion
* all predicates that are part of the join must be possible to be pushed down
  * predicates that cannot be represented as a series of AND or IN operations cannot be pushed down (e.g. LIKE or OR)
  * predicates that involve function calls
  * predicates that involve complex expressions like type coercion or addition/subtraction etc.
```

Or maybe we don't want to go so deep into implementation details. (I believe nobody should care if pushdown happened or not - if it can the engine will do it. What people actually care about is performance and they substitute pushdown as a proxy for performance).

posulliv replied:
I would prefer to link to a specific section that has more details on predicate pushdown rather than cover it here is part of join pushdown. What do you think?

mosabua replied:
Yeah ... we definitely need a separate section for predicate pushdown.. should we addressed in a separate PR though and then we can add a link to it here

mosabua replied:
There is a PR for predicate pushdown stuff in the works with @Jessie212

hashhar reviewed (commented):
Thanks a lot for contributing this. This looks good to me on the technical details % @mosabua's changes on wording.

This should serve as solid foundation to expand over time and add more details where needed. e.g. what types of predicates do certain connectors pushdown etc.

findinpath commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,65 @@ to perform the remaining query processing on a smaller amount of data.
+pushed down to the connector, the explain plan does **not** show that ``Join`` operator.
+The explain plan only shows the operations that are performed by Trino.
+
+As an example, we loaded the TPCH data set into a PostgreSQL database and then
```
I think that the documentation should tell the user how to setup himself a test postgres instance used to reproduce the pushdown join behavior.
I'd rather start from the supposition that the postgres database table is set and it contains two tables.

posulliv replied:
How to setup the TPCH  connector [is documented](https://trino.io/docs/current/connector/tpch.html) already. Configuring a Postgres connector is [also documented](https://trino.io/docs/current/connector/postgresql.html). Once you have those connectors configured, its just a `CREATE TABLE AS SELECT` statement to load TPCH date into a PostgreSQL database.

findinpath replied:
Sorry. I've meant to say that IMO the join push down should concentrate on what join push down is and not how to set up a postgres database in order to test it.

findinpath commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,65 @@ to perform the remaining query processing on a smaller amount of data.
+As an example, we loaded the TPCH data set into a PostgreSQL database and then
+queried it using the PostgreSQL connector::
+
+  SELECT c.custkey, o.orderkey
```
Please use `4` spaces to delimit the SQL blcok

posulliv commented:
Thanks for the comments @hashhar and @findinpath! Updated based on feedback.

findepi commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,65 @@ to perform the remaining query processing on a smaller amount of data.
+However, there are some generic conditions that must be met in order for a join
+to be pushed down:
+
+* join pushdown must be enabled
```
```suggestion
```

it's enabled by default, so this sentence is misleading -- user doesn't need to enable anything

while "join pushdown must not be disabled" would be correct thing to write, it would sound a bit silly to me. Maybe we need that still?

posulliv replied:
I believe join pushdown is disabled by default based on https://github.com/trinodb/trino/blob/ab9f02df0f5c2bf22ba12be6678c371dd91fa435/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadataConfig.java#L28-L34

This also aligns with what I have observed while using the feature in Trino 359.

findepi replied:
> I believe join pushdown is disabled by default based on

you're correct.

findepi commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,65 @@ to perform the remaining query processing on a smaller amount of data.
+
+* join pushdown must be enabled
+* all predicates that are part of the join must be possible to be pushed down
+* none of the tables in the join must be empty
```
```suggestion
```

where does this come from?
i am not aware of this condition

posulliv replied:
You are correct. I was looking at the wrong code for this condition.

findepi commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,65 @@ to perform the remaining query processing on a smaller amount of data.
+* join pushdown must be enabled
+* all predicates that are part of the join must be possible to be pushed down
+* none of the tables in the join must be empty
+* all tables in the join must be from the same catalog
```
A join is between two tables, not many.

```suggestion
* the tables in the join must be from the same catalog
```

findepi commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,65 @@ to perform the remaining query processing on a smaller amount of data.
+
+You can get the explain plan by prepending the above query with ``EXPLAIN``::
+
+    EXPLAIN
+    SELECT c.custkey, o.orderkey
```
```suggestion
    EXPLAIN SELECT c.custkey, o.orderkey
```

findepi commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,65 @@ to perform the remaining query processing on a smaller amount of data.
+
+.. caution::
+
+  Care should be taken when enabling join pushdown in a connector without statistics as
+  pushing down a join which can increase the row count compared to the size of
```
Join pushdown has nothing to do with statistics.

findepi commented:
@posulliv thanks for the update

posulliv commented:
@mosabua this is for ready for review if you get a chance. thanks!

mosabua commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,63 @@ to perform the remaining query processing on a smaller amount of data.
+* all predicates that are part of the join must be possible to be pushed down
+* the tables in the join must be from the same catalog
+
+You can check if pushdown for a specific join is performed by looking at the
```
You can verify

mosabua commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,63 @@ to perform the remaining query processing on a smaller amount of data.
+pushed down to the connector, the explain plan does **not** show that ``Join`` operator.
+The explain plan only shows the operations that are performed by Trino.
+
+As an example, we loaded the TPCH data set into a PostgreSQL database and then
```
Dont use "we" and dont use past tense "loaded" .. but this is all not relevant here really. I would remove the sentence.

mosabua commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,63 @@ to perform the remaining query processing on a smaller amount of data.
+
+You can check if pushdown for a specific join is performed by looking at the
+:doc:`EXPLAIN plan </sql/explain>` of the query. If a join is successfully
+pushed down to the connector, the explain plan does **not** show that ``Join`` operator.
```
Change to 

```
The explain plan does not show a ``Join`` operator, if the join is pushed down to the data source by the connector. 
```

mosabua commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,63 @@ to perform the remaining query processing on a smaller amount of data.
+You can check if pushdown for a specific join is performed by looking at the
+:doc:`EXPLAIN plan </sql/explain>` of the query. If a join is successfully
+pushed down to the connector, the explain plan does **not** show that ``Join`` operator.
+The explain plan only shows the operations that are performed by Trino.
```
remove this sentence

mosabua commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,63 @@ to perform the remaining query processing on a smaller amount of data.
+The explain plan only shows the operations that are performed by Trino.
+
+As an example, we loaded the TPCH data set into a PostgreSQL database and then
+queried it using the PostgreSQL connector::
```
I would remove this

mosabua commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,63 @@ to perform the remaining query processing on a smaller amount of data.
+    SELECT c.custkey, o.orderkey
+    FROM orders o JOIN customer c ON c.custkey = o.custkey;
+
+You can get the explain plan by prepending the above query with ``EXPLAIN``::
```
Prepend  ``EXPLAIN`` to a query to access the explain plan::

mosabua commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,63 @@ to perform the remaining query processing on a smaller amount of data.
+    EXPLAIN SELECT c.custkey, o.orderkey
+    FROM orders o JOIN customer c ON c.custkey = o.custkey;
+
+The explain plan for this query does not show any ``Join`` operator,
```
The following plan results from the PostgreSQL connector querying TPCH data in a PostgreSQL database. It does not show any ``Join`` operator as a result of the successful join push down::

mosabua commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,63 @@ to perform the remaining query processing on a smaller amount of data.
+
+.. caution::
+
+  Care should be taken when enabling join pushdown as pushing down a join
```
Sorry.. i fail to understand what this means. Why would a join push down result in more data transfer?

hashhar replied:
Some joins can be "expanding". e.g. Left side = 500MB, right side = 75MB but each row on right matches with 8 rows on the left side. So the size of the join can be 500 + (75*8) = 1.1GB which is actually larger than simply doing the join within Trino which is "just" 575MB. Almost a 50% savings.

mosabua replied:
hm .. so yes .. maybe the data transfer is larger .. but ultimately if the join is then run in Trino it would still amount to the same total data right? And the actual join operation is probably still faster in the underlying data source so I am not sure if you really need to think about this at all in the real world. I am more inclined to think that a join push down is always good even if it can result in a bit more data to transfer. Could it really incur large costs? Maybe some for data transfer .. but for performance?

mosabua replied:
At a minimum I would make this less prominent and forget about "caution" usage. Just a normal paragraph and a milder wording is good probably.

posulliv commented:
@mosabua thank you for the review and suggestions. I have applied all suggestions. Ready for another review whenever you get a chance.

posulliv commented:
@mosabua I converted the `caution` section to a regular paragraph and reworded it per feedback.

mosabua reviewed (approved):
Looks great now. Thanks!

hashhar commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,52 @@ to perform the remaining query processing on a smaller amount of data.
+However, there are some generic conditions that must be met in order for a join
+to be pushed down:
+
+* join pushdown must be enabled
```
Maybe mention the config property to be set?

hashhar replied:
Also, title case the bullet list?

posulliv replied:
I added the config property but didn't title case the bullet list to be consistent with the casing style used in other bullet lists in this section of the documentation.

hashhar reviewed (approved):
LGTM % comments.

electrum commented:
Nit: don't end commit message title in period. See https://chris.beams.io/posts/git-commit/

electrum commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,52 @@ to perform the remaining query processing on a smaller amount of data.
+However, there are some generic conditions that must be met in order for a join
+to be pushed down:
+
+* join pushdown must be enabled in a catalog using `join-pushdown.enabled`
```
This doesn't belong here, since this is for specific JDBC connectors

posulliv commented:
Thanks for the feedback. I updated the commit message and removed the reference to JDBC specific catalog property.

posulliv commented:
@electrum @mosabua Just checking in on this PR to see if there is anything else needed to get it merged?

electrum commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,52 @@ to perform the remaining query processing on a smaller amount of data.
+However, there are some generic conditions that must be met in order for a join
+to be pushed down:
+
+* join pushdown must be enabled in a catalog
```
This one is a bit strange, since this is a requirement for specific connectors that have a configuration flag to enable join pushdown. Also, I think the choice to disable by default for JDBC connectors is wrong -- I suspect that most joins do not increase the cardinality.

We could either

* remove this point and add a sentence below that says something like "A connector may require additional configuration to enable join pushdown."
* make it enabled by default for JDBC connectors and then there is no motivation to discuss this here at all

electrum commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,52 @@ to perform the remaining query processing on a smaller amount of data.
+* the tables in the join must be from the same catalog
+
+You can verify if pushdown for a specific join is performed by looking at the
+:doc:`EXPLAIN plan </sql/explain>` of the query. The explain plan does not
```
Nit: I would make this
```
:doc:`/sql/explain` plan
```
So just the `EXPLAIN` part is a link

electrum commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,52 @@ to perform the remaining query processing on a smaller amount of data.
+show a ``Join`` operator, if the join is pushed down to the data source by the
+connector.
+
+Prepend ``EXPLAIN`` to a query to access the explain plan::
```
Let's remove this sentence, since it's repeating what is in the EXPLAIN documentation linked above, and in the above sections. It certainly doesn't warrant a separate paragraph, and we don't need to repeat this for every section.

electrum commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -151,6 +151,52 @@ to perform the remaining query processing on a smaller amount of data.
+         orderkey := orderkey_0:bigint:int8
+         custkey := custkey_1:bigint:int8
+
+Pushing down a join can increase the row count compared to the size of the
```
This sentence is one sided. It can increase, but it can also decrease. The way this is written implies that it's a bad thing to do. Typically, it is a big win.

posulliv replied:
I made an attempt to clarify it. The only reason I put this here was to try and provide reasoning for why it was disabled by default in the JDBC connectors. If we enable join pushdown by default in JDBC connectors then. I think this sentence should just be removed altogether.

posulliv commented:
@electrum thanks for the comments. Updated again. I removed the sentence referring to enabling join pushdown. I can open a PR to enable join pushdown by default in the JDBC connectors if that makes sense to you.

hashhar commented on docs/src/main/sphinx/optimizer/pushdown.rst:
```diff
@@ -181,6 +181,50 @@ to perform the remaining query processing on a smaller amount of data.
+
+It is typically beneficial to push down a join. Pushing down a join can also
+increase the row count compared to the size of the input to the join. This
+may incur costs due to increased network traffic.
```
WDYT? `impact performance and incur costs` (or maybe leave out costs)

hashhar reviewed (approved):
Looks good to me % any comments from @electrum

hashhar commented:
This looks good to me. If there are any remaining concerns let's address them in follow-ups.
